### PR TITLE
OptimizationGradientError

### PR DESCRIPTION
**Pull request by jduerholt**
## Motivation

I am currently experimenting with some custom kernels, which exhibit some numeric instabilities, which lead in some situations during optimization to `NaN`s in the gradient. To better examine these situations, I thought it could be nice if the error which is thrown in this situation also contains the current x which leads to the `NaN`s in the gradient. For this purpose, I implemented an `OptimizationGradientError` like the `OptimizationTimeoutError` which holds the current x.

Im also thinking about catching the error in `timeout.py` as the `OptimizationTimeoutError`, throwing a warning and returning a result of the optimization. What do you think about this? I think this could be a useful and elegant way of dealing with it.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/pytorch/botorch/blob/main/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

Unit tests.

### CODE REVIEW

File: test/exceptions/test_errors.py [hunk 1]
```diff
@@ -49,3 +50,9 @@ def test_OptimizationTimeoutError(self):
         self.assertTrue(np.array_equal(error.current_x, np.array([1.0])))
         with self.assertRaises(OptimizationTimeoutError):
             raise error
+
+    def test_OptimizationGradientError(self):
+        error = OptimizationGradientError("message", current_x=np.array([1.0]))
+        self.assertTrue(np.array_equal(error.current_x, np.array([1.0])))
+        with self.assertRaises(OptimizationGradientError):
+            raise error
```
Comment (Balandat) [inline test/exceptions/test_errors.py:58]:
```suggestion
        with self.assertRaisesRegex(OptimizationGradientError, "message"):
            raise error
```

File: test/generation/test_gen.py [hunk 2]
```diff
@@ -318,7 +319,7 @@ def test_gen_candidates_scipy_nan_handling(self):
             test_grad = torch.tensor([0.5, 0.2, float("nan")], **ckwargs)
             # test NaN in grad
             with mock.patch("torch.autograd.grad", return_value=[test_grad]):
-                with self.assertRaisesRegex(RuntimeError, expected_regex):
+                with self.assertRaisesRegex(OptimizationGradientError, expected_regex):
```
Comment (Balandat) [inline test/generation/test_gen.py:322]: One thing to consider here is that this could break people's code if they rely on this being a `RuntimeError` - how about subclassing `OptimizationGradientError` from both `BoTorchError` and `RuntimeError`? (this is somehting we could probably consider doing more generally)

**Review by Balandat — COMMENTED** [review]
Thanks for putting this up - this makes sense to me. 

> Im also thinking about catching the error in timeout.py as the OptimizationTimeoutError, throwing a warning and returning a result of the optimization. What do you think about this? I think this could be a useful and elegant way of dealing with it.

Yeah I think this generally make sense. One issue with warnings is that they're often overlooked or ignored - so I wouldn't want to change the default behavior that something that would previously throw an error would now just return a result. But maybe it could be gated by some `raise_on_nan` flag?

File: test/generation/test_gen.py (hunk 2 above)
Comment (jduerholt) [reply test/generation/test_gen.py:322]: Good point, I will change it accordingly.

File: botorch/exceptions/errors.py [hunk 3]
```diff
@@ -72,3 +72,17 @@ def __init__(
         super().__init__(*args, **kwargs)
         self.current_x = current_x
         self.runtime = runtime
+
+
+class OptimizationGradientError(BotorchError):
```
Comment (Balandat) [inline botorch/exceptions/errors.py:77]:
```suggestion
class OptimizationGradientError(BotorchError, RuntimeError):
```

**Review by esantorella — APPROVED** [review]
Thanks! I'm all for making it easier to debug optimization errors :)

**Comment by jduerholt** [comment]
I think it is also fine as it is now, advanced users can now catch the error and see which is the `x` that creates the `NaN`s and further inspect them. From my side, no urgent need to further catch in `minimize_with_timeout`, based on an argument to the function. Maybe, I add this in a seperate PR, but from my side, this is good to go now. Thanks for the quick review!
